public/app/features/panel/panellinks/link_srv.ts

```diff
@@ -235,22 +235,26 @@ export interface LinkService {
   getDataLinkUIModel: <T>(link: DataLink, scopedVars: ScopedVars, origin: T) => LinkModel<T>;
 }
 
+export interface LinkServiceDependencies {
+  templateSrv: TemplateSrv;
+  timeSrv: () => TimeSrv;
+}
+
 export class LinkSrv implements LinkService {
-  /** @ngInject */
-  constructor(private templateSrv: TemplateSrv, private timeSrv: TimeSrv) {}
+  constructor(private dependencies: LinkServiceDependencies = { templateSrv: templateSrv, timeSrv: getTimeSrv }) {}
```

Comment:
What is the reason for having these arguments grouped into a `dependencies` object? To me it feels like any constructor argument can be seen as dependency so is this to mimic the React component Props somehow?

Reply:
It's so we can set default values for run-time code, does that answer your question?

Reply:
Not sure actually, couldn't we just do?:
```
constructor(private templateSrv = templateSrv, private timeSrv = getTimeSrv) {}
```
Or is that because we need to do something like?:
```
new LinkSrv(undefined, timeSrv);
```
I am not sure if passing undefined actually triggers the default value or not. But it based on the types here it does not seem we can provide just `timeSrv` and let `templateSrv` use default either.

Reply:
Yeah, I was wondering the same as @aocenas is. Can't see a huge benefit of `dependencies` object, but maybe I'm missing something here.

Reply:
I'm not an Angular expert, but wouldn't templateSrv/timeSrv in the constructor be resolved by Angular in that example?

Reply:
@hugohaggmark I think it wouldn't without `ngInject` annotation. Also, in order for Angular to pick it up it needs to be registered with `coreModule.service('linkSrv', LinkSrv);` which I think is not a case anymore with this proposal :)